### PR TITLE
feat: allow four digit versioning by way of revision field

### PR DESCRIPTION
**Pull request by mikefero**
This feature adds a four digit version parsing while still allowing strict semver. Kong Gateway and Kong Gateway OSS use different versioning systems; Kong Gateway doesn't follow strict semver. As Kong Gateway and Kong Gateway OSS are released the major.minor.patch versions are in lock step; however, Kong Gateway may have additional revisions that are based of the OSS upstream.

Comparison and version ranges can be mixed between 3 and 4 digit versions. When comparing 3 digit version to a 4 digit version only the major.minor.patch-prerelease+buildmetadata will apply. This allows the comparison of Kong Gateway and Kong Gateway OSS versions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,16 +12,16 @@ The current stable version is [*v4*](v4/) and is fully go-mod compatible.
 Usage
 -----
 ```bash
-$ go get github.com/blang/semver/v4
+$ go get github.com/mikefero/semver/v4
 # Or use fixed versions
-$ go get github.com/blang/semver/v4@v4.0.0
+$ go get github.com/mikefero/semver/v4@v4.0.0
```
Comment (javierguerragiraldez) [inline README.md]: i guess it's now
```suggestion
$ go get github.com/Kong/semver/v4@v4.0.0
```

File: package.json [hunk 2]
```diff
@@ -1,17 +1,20 @@
 {
   "author": "blang",
+  "maintainers": [
+    "mikefero"
+  ],
   "bugs": {
-    "URL": "https://github.com/blang/semver/issues",
-    "url": "https://github.com/blang/semver/issues"
+    "URL": "https://github.com/mikefero/semver/issues",
+    "url": "https://github.com/mikefero/semver/issues"
```
Comment (javierguerragiraldez) [inline package.json]: again `s/mikefero/Kong/g`

File: range.go [hunk 3]
```diff
@@ -301,6 +304,19 @@ func incrementMinorVersion(vStr string) (string, error) {
 	return strings.Join(parts, "."), nil
 }
 
+// incrementPatchVersion will increment the patch version
+// of the passed version
+func incrementPatchVersion(vStr string) (string, error) {
+	parts := strings.Split(vStr, ".")
+	i, err := strconv.Atoi(parts[2])
+	if err != nil {
+		return "", err
```
Comment (javierguerragiraldez) [inline range.go]: this error could be wrapped to give some context

File: range_test.go [hunk 4]
```diff
@@ -295,61 +345,125 @@ func TestExpandWildcardVersion(t *testing.T) {
...
 func TestVersionRangeToRange(t *testing.T) {
-	vr := versionRange{
-		v: MustParse("1.2.3"),
-		c: compLT,
-	}
-	rf := vr.rangeFunc()
-	if !rf(MustParse("1.2.2")) || rf(MustParse("1.2.3")) {
-		t.Errorf("Invalid conversion to range func")
-	}
+	t.Run("3 digit ranges", func(t *testing.T) {
```
Comment (javierguerragiraldez) [inline range_test.go:348]: just curious, does the use of `t.Run` raises the minimum Go version usable?

i know the Go community is quite fast to update the compiler, so it shouldn't be a problem, unless it's declared in some machine-readable file.

File: semver.go [hunk 5]
```diff
@@ -11,32 +12,78 @@ const (
 	numbers  string = "0123456789"
 	alphas          = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-"
 	alphanum        = alphas + numbers
+
+	// threeFourDigitVersionRegexFormat represents the strict rules for parsing 3 and 4 digit
+	// version numbers. There are six capture groups which define major, minor, patch, revision,
+	// pre-release, and build metadata. When using 3 digit versioning strict semver adherence
+	// will be utilized/
+	threeFourDigitVersionRegexFormat = `^(?P<major>0|[1-9]\d*)\.` +
```
Comment (javierguerragiraldez) [inline semver.go:20]: i'm never sure about operator precedence in regexes.  is `0|[1-9]\d*` equivalent to `(0)|([1-9]\d*)` or `(0|[1-9])\d*`? if the latter, it would allow numbers like `01`.

File: semver.go [hunk 6]
```diff
@@ -11,32 +12,78 @@ const (
...
+	Major    uint64
+	Minor    uint64
+	Patch    uint64
+	Revision int64
+	Pre      []PRVersion
+	Build    []string //No Precedence
 }
 
-// Version to string
+// KongVersion to string
```
Comment (javierguerragiraldez) [inline semver.go]: what is a `KongVersion`?  (and how is it kongverted ? :laughing:  )

File: semver.go [hunk 7]
```diff
@@ -64,12 +111,16 @@ func (v Version) String() string {
 // FinalizeVersion discards prerelease and build number and only returns
 // major, minor and patch number.
```
Comment (javierguerragiraldez) [inline semver.go]: ... and optionally revision

File: semver.go [hunk 8]
```diff
@@ -173,16 +234,30 @@ func (v Version) Compare(o Version) int {
 
 }
 
+// IncrementPatch increments the patch version
+func (v *Version) IncrementRevision() error {
+	if v.Revision >= 0 {
+		v.Revision++
+	}
+	return nil
```
Comment (javierguerragiraldez) [inline semver.go:246]: this will silently be ignored if there wasn't a revision number.  i'd guess calling this should count as "first revision".  otherwise, return an error.

or am i misinterpreting the semantics?   if revision>=0 means "i do use revisions" while revision==-1 means "this is a real semver", then this function should fail on revision==-1.

File: semver.go [hunk 9]
```diff
@@ -238,31 +316,74 @@ func Make(s string) (Version, error) {
...
-	// Remove leading zeros.
-	for i, p := range parts {
-		if len(p) > 1 {
-			p = strings.TrimLeft(p, "0")
-			if len(p) == 0 || !strings.ContainsAny(p[0:1], "0123456789") {
-				p = "0" + p
-			}
-			parts[i] = p
+	if !TolerantVersionRegex.MatchString(s) {
+		return Version{}, fmt.Errorf("Invalid tolerant version: '%s'", s)
```
Comment (javierguerragiraldez) [inline semver.go]: nit: the "tolerant" word sounds weird here, as it's not the version the tolerant here, it's the parser.  and i don't think it adds any meaning.  not for the developer (who knows which parser is being used) nor for the user (who shouldn't care).

(subnit: thou shall not capitalise errors)

File: semver.go [hunk 10]
```diff
@@ -238,31 +316,74 @@ func Make(s string) (Version, error) {
...
+	s = fmt.Sprintf("%s.%s.%s", majorStr, minorStr, patchStr)
+	if len(revisionStr) > 0 {
+		s = fmt.Sprintf("%s.%s", s, revisionStr)
+	}
+	if len(prereleaseStr) > 0 {
+		s = fmt.Sprintf("%s-%s", s, prereleaseStr)
+	}
+	if len(buildStr) > 0 {
+		s = fmt.Sprintf("%s+%s", s, buildStr)
+	}
```
Comment (javierguerragiraldez) [inline semver.go:395]: what about `strings.Builder`?

File: package.json (hunk 2 above)
Comment (mikefero) [reply package.json]: That's what I get for starting on my fork ;).

File: range_test.go (hunk 4 above)
Comment (mikefero) [reply range_test.go:348]: I am not sure, I will have to look into that one, the version listed in go.mod is 1.12 if I remember correctly.

File: semver.go (hunk 5 above)
Comment (mikefero) [reply semver.go:20]: So I modified this from https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string adding the revision and created the tolerant variant too.

File: semver.go (hunk 9 above)
Comment (mikefero) [reply semver.go]: Yeah I don't usually capitalize errors, but this module is riddled with bad practices so I kept things this way to be consistent.

File: semver.go (hunk 8 above)
Comment (mikefero) [reply semver.go:246]: So `-1` means revision is disabled and there should be no changes to revision if you call it; if there is a revision >=0 then increment. If revision is not parsed then we should make sure it never gets incremented or else it could get added to the string.

File: semver.go (hunk 5 above)
Comment (javierguerragiraldez) [reply semver.go:20]: ok, so i tried it: https://go.dev/play/p/ANozC6L9sJM with surprising results:
- `^0|[1-9]\d*$` accepts "01".  not good
- `^0|([1-9]\d*)$` also accepts "01".... why?
- `^(0|[1-9]\d*)$`  does reject "01"  so it seems these are fine.

found the explanation in [a POSIX doc](https://www.boost.org/doc/libs/1_56_0/libs/regex/doc/html/boost_regex/syntax/basic_extended.html#boost_regex.syntax.basic_extended.operator_precedence).   in short: alternation '|' is the lowest precedence of all, which is good, but it's lower than anchors (^ and $).

the second pattern essentially means "starts with 0 or ends with digits", which isn't useful.  the last pattern is basically: "full string: 0 or nice number"

TL;DR: these patterns are ok because each alternation is wrapped in some group

File: semver.go [hunk 11]
```diff
@@ -11,32 +12,78 @@ const (
...
+	// threeFourDigitVersionRegexFormat represents the strict rules for parsing 3 and 4 digit
+	// version numbers. There are six capture groups which define major, minor, patch, revision,
+	// pre-release, and build metadata. When using 3 digit versioning strict semver adherence
+	// will be utilized/
+	threeFourDigitVersionRegexFormat = `^(?P<major>0|[1-9]\d*)\.` +
+		`(?P<minor>0|[1-9]\d*)\.` +
+		`(?P<patch>0|[1-9]\d*)` +
+		`(?:\.(?P<revision>0|[1-9]\d*))?` +
+		`(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?` +
+		`(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`
```
Comment (GGabriele) [inline semver.go:25]: Mother of regex :D

File: semver.go [hunk 12]
```diff
@@ -11,32 +12,78 @@ const (
...
+
+	// threeFourDigitTolerantVersionRegexFormat represents loose rules for parsing 3 and 4 digit
+	// version numbers. There are six capture groups which define major, minor, patch, revision,
+	// pre-release, and build metadata. The loose format allows for leading and trailing spaces,
+	// prefixing versions with a 'v' and empty capture groups in order to dynamically fill in the
+	// fields with defaults. When using 3 digit versioning loose semver adherence will be utilized.
+	threeFourDigitTolerantVersionRegexFormat = `^(?:\s*)?(?:v)?` +
```
Comment (GGabriele) [inline semver.go]: Maybe using `Loose` instead of `Tolerant`? The latter doesn't sound right, but maybe it's because of my not-native-english ears :P

Comment (mikefero) [reply semver.go]: Indeed. I chose tolerant because of the `ParseTolerant()` method call, but I think that makes more sense as the function name and action that it performs; loose or some form of that does sound more appealing.

File: range.go (hunk 3 above)
Comment (mikefero) [reply range.go]: Wrapped this and all the other increment functions in range.go

File: range_test.go (hunk 4 above)
Comment (mikefero) [reply range_test.go:348]: Looks like it was added in [go 1.7](https://pkg.go.dev/testing#T.Run) so we should be good here; good thinking.

File: semver.go (hunk 6 above)
Comment (mikefero) [reply semver.go]: It is all the rage and it converts to to all things; that's just how it rolls.

File: semver.go (hunk 12 above)
Comment (mikefero) [reply semver.go]: Updating the name to permissive and also updating the documentation to give more context.

File: semver.go (hunk 9 above)
Comment (mikefero) [reply semver.go]: Updating the name of the regex to use the word "permissive" and also updating the poorly written log message by removing `tolerant`.

File: semver.go (hunk 8 above)
Comment (mikefero) [reply semver.go:246]: I updated the godoc for this function and made it a little more descriptive; let me know if this clears things up.

File: semver.go [hunk 13]
```diff
@@ -374,18 +379,26 @@ func ParseTolerant(s string) (Version, error) {
 	}
 
 	// Generate version to properly parse
-	s = fmt.Sprintf("%s.%s.%s", majorStr, minorStr, patchStr)
+	b := strings.Builder{}
+	b.WriteString(majorStr)
+	b.WriteString(".")
+	b.WriteString(minorStr)
+	b.WriteString(".")
+	b.WriteString(patchStr)
```
Comment (javierguerragiraldez) [inline semver.go]:
```suggestion
	fmt.Fprintf(&b, "%s.%s.%s", majorStr, minorStr, patchStr)
```

File: semver.go [hunk 14]
```diff
@@ -374,18 +379,26 @@ func ParseTolerant(s string) (Version, error) {
...
 	if len(revisionStr) > 0 {
-		s = fmt.Sprintf("%s.%s", s, revisionStr)
+		b.WriteString(".")
+		b.WriteString(revisionStr)
```
Comment (javierguerragiraldez) [inline semver.go:386]:
```suggestion
		fmt.Fprintf(&b, "%s.%s", s, revisionStr)
```

Comment (mikefero) [reply semver.go:386]: This is adding the full version string back in to the string builder; pretty sure we'll need to keep the `WriteString` calls here to build the rest of the string.

Comment (javierguerragiraldez) [reply semver.go:386]: oh, right, it should be `fmt.Fprintf(&b, ".%s", revisionStr)`

but of course, to replace just two `WriteString`s it's not that much of an improvement

**Comment by mikefero** [comment]
I am going to remove Travis here as well and make a very simplistic workflow for GHA.
